# Imports

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
raw_data = pd.read_csv('hotel_bookings.csv')
raw_data.sample(8)

hotel  is_canceled  lead_time  arrival_date_year  \
23624   Resort Hotel            0         12               2016   
74205     City Hotel            1        363               2015   
26620   Resort Hotel            0          1               2016   
110400    City Hotel            0          4               2017   
23148   Resort Hotel            0          5               2016   
64674     City Hotel            1         37               2017   
62194     City Hotel            1         86               2017   
86413     City Hotel            0          4               2016   

       arrival_date_month  arrival_date_week_number  \
23624               April                        18   
74205             October                        42   
26620              August                        32   
110400              April                        17   
23148               April                        15   
64674               March                        11   
62194             January                         1   
86413               April                        14   

        arrival_date_day_of_month  stays_in_weekend_nights  \
23624                          27                        0   
74205                          15                        0   
26620                           5                        0   
110400                         27                        0   
23148                           9                        2   
64674                          15                        0   
62194                           4                        0   
86413                           2                        0   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
23624                      3       2  ...    No Deposit  385.0     NaN   
74205                      2       2  ...    Non Refund    1.0     NaN   
26620                      1       1  ...    No Deposit    NaN     NaN   
110400                     3       2  ...    No Deposit    9.0     NaN   
23148                      5       1  ...    No Deposit   40.0     NaN   
64674                      2       2  ...    Non Refund  378.0     NaN   
62194                      3       2  ...    No Deposit    9.0     NaN   
86413                      1       2  ...    No Deposit    9.0     NaN   

       days_in_waiting_list    customer_type     adr  \
23624                     0  Transient-Party   85.33   
74205                     0         Contract   62.00   
26620                     0        Transient  172.00   
110400                    0        Transient    0.00   
23148                     0         Contract   51.00   
64674                     0        Transient  110.00   
62194                     0        Transient   79.20   
86413                     0        Transient   99.00   

        required_car_parking_spaces  total_of_special_requests  \
23624                             0                          0   
74205                             0                          0   
26620                             0                          0   
110400                            0                          2   
23148                             0                          0   
64674                             0                          0   
62194                             0                          2   
86413                             0                          1   

        reservation_status reservation_status_date  
23624            Check-Out                30-04-16  
74205             Canceled                01-01-15  
26620            Check-Out                06-08-16  
110400           Check-Out                30-04-17  
23148            Check-Out                16-04-16  
64674             Canceled                06-02-17  
62194             Canceled                12-12-16  
86413            Check-Out                03-04-16  

[8 rows x 32 columns]

## 1. Data Collection


In [3]:
raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

- There are 32 columns in the dataset.
- The dataset contains 119,390 rows.
- Data Types:
    - 1. 16 columns are of int64 data type.
    - 2. 4 columns are of float64 data type.
    - 3. 12 columns are of object data types. 
- Missing Values:
    - The column `children` has <span style="color:red">4</span> missing values.
    - The column `country` has <span style="color:red">488</span> missing values.
    - The column `agent` has <span style="color:red">16,340</span> missing values.
    - The column `company` has <span style="color:red">112,593</span> missing values.


## 2 Exploratory Data Analysis (EDA)

### 2.1 Looking for missing values.

In [4]:
raw_data.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

- Dropping null values is easy but isn't a good practice, so we will start with replacing them with zero.

In [5]:
raw_data.fillna(0, inplace=True)

### 2.2 cleaning Data

In [6]:
raw_data.describe().T

count         mean         std      min  \
is_canceled                     119390.0     0.370416    0.482918     0.00   
lead_time                       119390.0   104.011416  106.863097     0.00   
arrival_date_year               119390.0  2016.156554    0.707476  2015.00   
arrival_date_week_number        119390.0    27.165173   13.605138     1.00   
arrival_date_day_of_month       119390.0    15.798241    8.780829     1.00   
stays_in_weekend_nights         119390.0     0.927599    0.998613     0.00   
stays_in_week_nights            119390.0     2.500302    1.908286     0.00   
adults                          119390.0     1.856403    0.579261     0.00   
children                        119390.0     0.103886    0.398555     0.00   
babies                          119390.0     0.007949    0.097436     0.00   
is_repeated_guest               119390.0     0.031912    0.175767     0.00   
previous_cancellations          119390.0     0.087118    0.844336     0.00   
previous_bookings_not_canceled  119390.0     0.137097    1.497437     0.00   
booking_changes                 119390.0     0.221124    0.652306     0.00   
agent                           119390.0    74.828319  107.141953     0.00   
company                         119390.0    10.775157   53.943884     0.00   
days_in_waiting_list            119390.0     2.321149   17.594721     0.00   
adr                             119390.0   101.831122   50.535790    -6.38   
required_car_parking_spaces     119390.0     0.062518    0.245291     0.00   
total_of_special_requests       119390.0     0.571363    0.792798     0.00   

                                    25%       50%     75%     max  
is_canceled                        0.00     0.000     1.0     1.0  
lead_time                         18.00    69.000   160.0   737.0  
arrival_date_year               2016.00  2016.000  2017.0  2017.0  
arrival_date_week_number          16.00    28.000    38.0    53.0  
arrival_date_day_of_month          8.00    16.000    23.0    31.0  
stays_in_weekend_nights            0.00     1.000     2.0    19.0  
stays_in_week_nights               1.00     2.000     3.0    50.0  
adults                             2.00     2.000     2.0    55.0  
children                           0.00     0.000     0.0    10.0  
babies                             0.00     0.000     0.0    10.0  
is_repeated_guest                  0.00     0.000     0.0     1.0  
previous_cancellations             0.00     0.000     0.0    26.0  
previous_bookings_not_canceled     0.00     0.000     0.0    72.0  
booking_changes                    0.00     0.000     0.0    21.0  
agent                              7.00     9.000   152.0   535.0  
company                            0.00     0.000     0.0   543.0  
days_in_waiting_list               0.00     0.000     0.0   391.0  
adr                               69.29    94.575   126.0  5400.0  
required_car_parking_spaces        0.00     0.000     0.0     8.0  
total_of_special_requests          0.00     0.000     1.0     5.0

In [7]:
# Display rows with negative adr values
negative_adr_rows = raw_data[raw_data['adr'] < 0]
display(negative_adr_rows)

# Drop rows with negative adr values
raw_data.drop(negative_adr_rows.index, inplace=True)

hotel  is_canceled  lead_time  arrival_date_year  \
14969  Resort Hotel            0        195               2017   

      arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
14969              March                        10                          5   

       stays_in_weekend_nights  stays_in_week_nights  adults  ...  \
14969                        4                     6       2  ...   

       deposit_type  agent company days_in_waiting_list    customer_type  \
14969    No Deposit  273.0     0.0                    0  Transient-Party   

        adr  required_car_parking_spaces  total_of_special_requests  \
14969 -6.38                            0                          0   

       reservation_status reservation_status_date  
14969           Check-Out                15-03-17  

[1 rows x 32 columns]

### Based on the following logic:

1.  Rows containing negative values in the `adr` column.
2.  The `distribution_channel` column, as it depicts the same information as the `market_segment` column.
3.  The `arrival_date_week_number` column, as the week number can be derived from the `arrival_date` column.
4.  Rows where both the `children` and `adults` columns contain zero.

In [8]:
def clean_dataframe(data):
    """
    Cleans a pandas DataFrame based on specified criteria.

    Args:
        raw_data: The pandas DataFrame to clean.

    Returns:
        The cleaned pandas DataFrame.
    """

    # 1. Drop rows with negative 'adr' values.
    data.drop(data[data['adr'] < 0].index, inplace=True)

    # 2. Drop the 'distribution_channel' column.
    if 'distribution_channel' in data.columns: # check if column exists
        data = data.drop(columns=['distribution_channel'])

    # 3. Drop the 'arrival_date_week_number' column.
    if 'arrival_date_week_number' in data.columns: # check if column exists
        data = data.drop(columns=['arrival_date_week_number'])

    # 4. Drop rows where both 'children' and 'adults' are zero.
    data = data[~((data['children'] == 0) & (data['adults'] == 0))] # ~ is the not operator.

    return data


raw_data = clean_dataframe(raw_data)

In [9]:
raw_data

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_day_of_month  stays_in_weekend_nights  \
0                    July                          1                        0   
1                    July                          1                        0   
2                    July                          1                        0   
3                    July                          1                        0   
4                    July                          1                        0   
...                   ...                        ...                      ...   
119385             August                         30                        2   
119386             August                         31                        2   
119387             August                         31                        2   
119388             August                         31                        2   
119389             August                         29                        2   

        stays_in_week_nights  adults  children  ...  deposit_type  agent  \
0                          0       2       0.0  ...    No Deposit    0.0   
1                          0       2       0.0  ...    No Deposit    0.0   
2                          1       1       0.0  ...    No Deposit    0.0   
3                          1       1       0.0  ...    No Deposit  304.0   
4                          2       2       0.0  ...    No Deposit  240.0   
...                      ...     ...       ...  ...           ...    ...   
119385                     5       2       0.0  ...    No Deposit  394.0   
119386                     5       3       0.0  ...    No Deposit    9.0   
119387                     5       2       0.0  ...    No Deposit    9.0   
119388                     5       2       0.0  ...    No Deposit   89.0   
119389                     7       2       0.0  ...    No Deposit    9.0   

       company days_in_waiting_list  customer_type     adr  \
0          0.0                    0      Transient    0.00   
1          0.0                    0      Transient    0.00   
2          0.0                    0      Transient   75.00   
3          0.0                    0      Transient   75.00   
4          0.0                    0      Transient   98.00   
...        ...                  ...            ...     ...   
119385     0.0                    0      Transient   96.14   
119386     0.0                    0      Transient  225.43   
119387     0.0                    0      Transient  157.71   
119388     0.0                    0      Transient  104.40   
119389     0.0                    0      Transient  151.20   

        required_car_parking_spaces total_of_special_requests  \
0                                 0                         0   
1                                 0                         0   
2                                 0                         0   
3                                 0                         0   
4                                 0                         1   
...                             ...                       ...   
119385                            0                         0   
119386                            0                        

### 2.3 Feature Selection

#### Why Feature Selection Might Be Beneficial:

* **Improved LLM Performance:** LLMs can be sensitive to irrelevant or redundant information. Feature selection can help focus the LLM on the most important data points, potentially improving accuracy and relevance.

* **Reduced Dimensionality:** If your dataset has a large number of features, feature selection can reduce the dimensionality of the data. This can lead to:

    1.  Faster vector embedding generation and retrieval.
    2.  Reduced storage requirements for your vector database.
    3.  Improved efficiency of your API.